src/Generating/PieceModifier.cpp

```diff
@@ -0,0 +1,410 @@
+
+// PieceModifier.cpp
+
+// Implements the various classes descending from cPiece::cPieceModifier
+
+#include "Globals.h"
+#include "PieceModifier.h"
+#include "../Noise/Noise.h"
+
+
+
+
+
+// Constant that is added to seed
+static const int SEED_OFFSET = 135 * 13;
+
+
+
+
+
+// Emit a warning if the first param is true
+#define CONDWARNING(ShouldLog, Fmt, ...) \
+	do { \
+		if (ShouldLog) \
+		{ \
+			LOGWARNING(Fmt, __VA_ARGS__); \
+		} \
+	} while (false)
+
+
+
+
+
+////////////////////////////////////////////////////////////////////////////////
+/** A modifier which is pseudo-randomly replacing blocks to other types and metas. */
+class cPieceModifierRandomizeBlocks:
+	public cPiece::cPieceModifier
+{
+public:
+	cPieceModifierRandomizeBlocks(void) :
+		m_Seed()
+	{
+	}
+
+	virtual bool InitializeFromString(const AString & a_Params, bool a_LogWarnings) override
+	{
+		AString Params = a_Params;
+
+
+		/** BlocksToReplace parsing */
+		auto idxPipe = Params.find('|');
+		if (idxPipe != AString::npos)
+		{
+			AString blocksToReplaceStr = Params.substr(0, idxPipe);
+			auto blocksToReplace = StringSplitAndTrim(blocksToReplaceStr, ",");
+			for (size_t i = 0; i < blocksToReplace.size(); i++)
+			{
+				BLOCKTYPE blockType = static_cast<BLOCKTYPE>(BlockStringToType(blocksToReplace[i]));
+				if ((blockType == E_BLOCK_AIR) && !NoCaseCompare(blocksToReplace[i], "Air"))
+				{
+					CONDWARNING(a_LogWarnings, "Cannot parse block type from string \"%s\"!", blocksToReplace[i].c_str());
+					return false;
+				}
+				m_BlocksToReplace[blockType] = 1;
+			}
+
+			if (m_BlocksToReplace.size() == 0)
```

Comment:
Should this if condition be outside the `if (idxPipe != AString::npos)` scope? Otherwise it would still be possible to continue while having no pipeline right?